### PR TITLE
Make http read buffer retriable

### LINKED ISSUE #29696: Allow to setup retries in `URL` storage.
**Use case**

This query should successfully finish ten million HTTP requests in around a day:

```
SELECT * FROM url('https://hacker-news.firebaseio.com/v0/item/{1..10000000}.json', JSONEachRow,
$$
id UInt32,
deleted UInt8,
type Enum('story' = 1, 'comment' = 2, 'poll' = 3, 'pollopt' = 4, 'job' = 5),
by LowCardinality(String),
time DateTime,
text String,
dead UInt8,
parent UInt32,
poll UInt32,
kids Array(UInt32),
url String,
score Int32,
title String,
parts Array(UInt32),
descendants Int32
$$)
```

**Describe the solution you'd like**

If an error happened before any data has been consumed, we can retry a request.
Let's introduce a set of settings: `storage_url_max_retries`, `storage_url_retry_initial_backoff_ms`, `storage_url_retry_max_backoff_ms`. It will do up to max retries, starting from initial backoff ms, increasing 2 times on every retry but saturating at max backoff ms.

### PR DESCRIPTION
**Pull request by kssenii**
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Make reading from HTTP retriable. Closes https://github.com/ClickHouse/ClickHouse/issues/29696.

### CODE REVIEW

**Comment by tavplubix** [comment]
Also tests are required

**Comment by tavplubix** [comment]
Looks like retries work fine if an error happened in the middle of reading:
```
2021.11.09 23:09:02.389126 [ 200 ] {3fd05f32-450b-44fe-9ede-0727dcb09199} <Error> ReadWriteBufferFromHTTP: HTTP request to `http://minio1:9001/root/test_url_reconnect_505.tsv` failed at try 0/10 with bytes read: 0/unknown. Error: Timeout: connect timed out: 172.26.0.5:9001. (Current backoff wait is 100/2000 ms)
2021.11.09 23:09:02.492753 [ 200 ] {3fd05f32-450b-44fe-9ede-0727dcb09199} <Error> ReadWriteBufferFromHTTP: HTTP request to `http://minio1:9001/root/test_url_reconnect_505.tsv` failed at try 1/10 with bytes read: 0/517905509. Error: DB::Exception: Cannot read from istream at offset 0. (Current backoff wait is 200/2000 ms)
2021.11.09 23:09:02.741121 [ 200 ] {3fd05f32-450b-44fe-9ede-0727dcb09199} <Error> ReadWriteBufferFromHTTP: HTTP request to `http://minio1:9001/root/test_url_reconnect_505.tsv` failed at try 0/10 with bytes read: 5242880/517905509. Error: DB::Exception: Cannot read from istream at offset 6291456. (Current backoff wait is 100/2000 ms)
2021.11.09 23:09:02.863553 [ 200 ] {3fd05f32-450b-44fe-9ede-0727dcb09199} <Error> ReadWriteBufferFromHTTP: HTTP request to `http://minio1:9001/root/test_url_reconnect_505.tsv` failed at try 0/10 with bytes read: 7340032/517905509. Error: DB::Exception: Cannot read from istream at offset 3145728. (Current backoff wait is 100/2000 ms)
2021.11.09 23:09:02.966716 [ 200 ] {3fd05f32-450b-44fe-9ede-0727dcb09199} <Error> ReadWriteBufferFromHTTP: HTTP request to `http://minio1:9001/root/test_url_reconnect_505.tsv` failed at try 1/10 with bytes read: 7340032/517905509. Error: DB::Exception: Cannot read from istream at offset 0. (Current backoff wait is 200/2000 ms)
2021.11.09 23:09:04.170830 [ 200 ] {3fd05f32-450b-44fe-9ede-0727dcb09199} <Error> ReadWriteBufferFromHTTP: HTTP request to `http://minio1:9001/root/test_url_reconnect_505.tsv` failed at try 2/10 with bytes read: 7340032/517905509. Error: DB::Exception: Cannot read from istream at offset 0. (Current backoff wait is 400/2000 ms)
2021.11.09 23:09:04.584053 [ 200 ] {3fd05f32-450b-44fe-9ede-0727dcb09199} <Error> ReadWriteBufferFromHTTP: HTTP request to `http://minio1:9001/root/test_url_reconnect_505.tsv` failed at try 0/10 with bytes read: 8388608/517905509. Error: DB::Exception: Cannot read from istream at offset 2097152. (Current backoff wait is 100/2000 ms)
2021.11.09 23:09:05.687237 [ 200 ] {3fd05f32-450b-44fe-9ede-0727dcb09199} <Error> ReadWriteBufferFromHTTP: HTTP request to `http://minio1:9001/root/test_url_reconnect_505.tsv` failed at try 1/10 with bytes read: 8388608/517905509. Error: DB::Exception: Cannot read from istream at offset 0. (Current backoff wait is 200/2000 ms)
2021.11.09 23:09:06.889481 [ 200 ] {3fd05f32-450b-44fe-9ede-0727dcb09199} <Error> ReadWriteBufferFromHTTP: HTTP request to `http://minio1:9001/root/test_url_reconnect_505.tsv` failed at try 2/10 with bytes read: 8388608/517905509. Error: Timeout: connect timed out: 172.26.0.5:9001. (Current backoff wait is 400/2000 ms)
```

However, I'm not sure if it's expected or not, that number of attempt is not always increasing.

**Comment by kssenii** [comment]
Lets leave as is because https://github.com/ClickHouse/ClickHouse/issues/29696#issuecomment-964633219

File: src/IO/ReadWriteBufferFromHTTP.h [hunk 1]
```diff
@@ -222,41 +297,110 @@ namespace detail
...
+            return true;
         }
 
         bool nextImpl() override
         {
             if (next_callback)
                 next_callback(count());
 
-            if (use_external_buffer)
+            if (read_range.end && bytes_read == read_range.end.value())
```
Comment (tavplubix) [inline src/IO/ReadWriteBufferFromHTTP.h:309]: Will this condition work correctly if we initially have some offset, so number of bytes read is actually less than content length?

File: src/IO/ReadWriteBufferFromHTTP.h [hunk 2]
```diff
@@ -222,41 +297,110 @@ namespace detail
...
-                if (!working_buffer.empty())
-                    impl->position() = position();
+                try
+                {
+                    if (!impl)
+                    {
+                        /// If error is not retriable -- false is returned and exception is set.
+                        /// Otherwise the error is thrown and retries continue.
+                        bool initialized = initialize();
+                        if (!initialized && exception)
```
Comment (tavplubix) [inline src/IO/ReadWriteBufferFromHTTP.h:350]: Is it possible that `initialize()` returned false, but `exception` is `nullptr`? Or, for example, that `initialize` returned `true`, but `impl` is `nullptr`? Let's either remove unnecessary flags or add assertions.

File: src/IO/ReadWriteBufferFromHTTP.h [hunk 3]
```diff
@@ -117,14 +148,23 @@ namespace detail
...
                 request.set(std::get<0>(http_header_entry), std::get<1>(http_header_entry));
+
+            if (withPartialContent())
+            {
+                String range_header_value;
+                if (read_range.end)
+                    range_header_value = fmt::format("bytes={}-{}", read_range.begin + bytes_read, *read_range.end);
+                else
+                    range_header_value = fmt::format("bytes={}-", read_range.begin + bytes_read);
+                LOG_TRACE(log, "Adding header: Range: {}", range_header_value);
```
Comment (tavplubix) [inline src/IO/ReadWriteBufferFromHTTP.h:160]:
```suggestion
                LOG_TEST(log, "Adding header: Range: {}", range_header_value);
```
(I forgot to replace 'trace' with 'test' before committing the test)

**Review by tavplubix — APPROVED** [review]

**Pull request merged**